File: flang/test/OmpSs-RT/07_phases_ompss.dg/fortran/success_task_25.f90
! RUN: %oss-compile-and-run

! <testinfo>
! test_generator=(config/mercurium-ompss "config/mercurium-ompss-2 openmp-compatibility")
! </testinfo>

MODULE M
IMPLICIT NONE

TYPE :: T
   INTEGER :: VAR
END TYPE T

TYPE(T) :: X
CONTAINS
    SUBROUTINE SUB1()
        IMPLICIT NONE
        INTERFACE
           !$OSS TASK INOUT(X)
           SUBROUTINE SUB2(X)
              IMPORT T
              IMPLICIT NONE
              TYPE(T) :: X
           END SUBROUTINE SUB2
        END INTERFACE

       CALL SUB2(X)
    END SUBROUTINE SUB1
END MODULE M

SUBROUTINE SUB2(X)
   USE M, ONLY : T
   IMPLICIT NONE
   TYPE(T) :: X

   X % VAR = 1
END SUBROUTINE SUB2

PROGRAM MAIN
    USE M
    IMPLICIT NONE
    X % VAR = -1
    CALL SUB1()
    !$OSS TASKWAIT
    IF (X % VAR /= 1) THEN
        STOP 1
    ENDIF
END PROGRAM

File: clang/test/OmpSs-RT/07_phases_ompss.dg/c/success_task_25.c
/*--------------------------------------------------------------------
  (C) Copyright 2006-2013 Barcelona Supercomputing Center
                          Centro Nacional de Supercomputacion
  
  This file is part of Mercurium C/C++ source-to-source compiler.
  
  See AUTHORS file in the top level directory for information
  regarding developers and contributors.
  
  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 3 of the License, or (at your option) any later version.
  
  Mercurium C/C++ source-to-source compiler is distributed in the hope
  that it will be useful, but WITHOUT ANY WARRANTY; without even the
  implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
  PURPOSE.  See the GNU Lesser General Public License for more
  details.
  
  You should have received a copy of the GNU Lesser General Public
  License along with Mercurium C/C++ source-to-source compiler; if
  not, write to the Free Software Foundation, Inc., 675 Mass Ave,
  Cambridge, MA 02139, USA.
--------------------------------------------------------------------*/

// RUN: %oss-compile-and-run
// RUN: %oss-O2-compile-and-run

/*
<testinfo>
test_generator=(config/mercurium-ompss "config/mercurium-ompss-2 openmp-compatibility")
</testinfo>
*/
#include<assert.h>

int x;
int *v;

#pragma oss task inout(v[x])
void f()
{
    assert(v[x] == 1);
    v[x]--;
}

int main()
{
    x = 0;
    int a = 1;
    v = &a;

    f();

    #pragma oss task inout(v[x])
    {
        assert(v[x] == 0);
        v[x]++;
    }
    #pragma oss taskwait
    assert(v[x] == 1);
    return 0;
}
